Frontend/Testing.md

```diff
@@ -0,0 +1,336 @@
+    You usually check if for specific input, to a function for example, you get specific output.
+    Or if function returns nothing, you might test if something else was done inside its body like other
+    function call or some variable being modified.
+    Unit testing usually do developers who wrote the code.
+
+    One thing to remember is that you do not test code that is it not yours.
+    You do not test if React will call componentDidMount or not, or that JSX will be compiled correctly,
+    or that some external function will do whatever it should be doing.
+    External libraries and code should already have their tests.
+    This will be emphasized again later.
+
+2. Integration testing
+
+    Since we know what are unit tests, we can now describe what are integration tests and what is
+    integration testing.
+    Basically, it is testing of units combined and tested as group.
+    Purpose of these tests are checking if interaction between units is correct and that there are no
+    faults.
+    
+    It is more complex than unit testing, and sometimes you also need to have some configuration.
+    For example if you test that after some „add“ button click, your list will have one item more.
+    These tests usually do code authors as well.
+
+3. System testing
+
+    This kind of testing goes after implementation and in our domain it is tested in a browser.
+
+    You run the application and check how it is behaving.
+    For example, you open application and click button that should open modal, and modal should
+    have some different button that should do something else and so on.
+    This part can be automatized by tests which does this for you: clicking on a UI, expecting components
+    to be shown with correct data, typing in fields etc.
+    This type of tests can be done by other testers.
+
+    You can come across system testing being similar with end to end testing (often written as e2e).
+    We will give one point of view since this is quite debatable.
+    Let's say we have application that has list of items on homepage and next to each item there is add
+    to cart action.
+    After adding at least 1 item, going to cart is available.
+    In the cart there is summary of selected items and action to purchase it which goes to payment
+    service and after payment is done it should return to home page.
+    System testing will test if we can add items, go to cart, go to purchase and go back.
+    End to end testing will be testing same as system testing, but additionally, our balance should be
+    lowered for correct amount, and correct amount should be added to seller, correct items should be
+    ordered and so on.
+    So like a real world application usage.
+    As you can already see, end to end testing is really difficult to be done with automatic tests.
+    This kind of tests usually do QA people since it might require separate database and backend from staging/development one.
```

Comment:
```suggestion
    This kind of tests is usually done by QA people, since it might require a separate database and backend from staging/development one.
```